### PR TITLE
process bundle deletion events

### PR DESCRIPTION
**Pull request by MDunitz**
## Need
- to safely delete bundles and their associated bundle_file_links and any files that were only linked to that bundle when receiving a bundle deletion event from dss

## Approach
- delete bundle_file_links and get a list of file_fqids
- delete any files that arent associated with any other bundles
- delete the bundle

## Tests
- added unit tests
- tested directly on db

## Todo
- once we include file version/fqid in the process_file_link table also delete associated processes

### CODE REVIEW

File: dcpquery/etl/__init__.py [hunk 1]
```diff
@@ -255,3 +275,12 @@ def process_bundle_event(dss_event):
         drop_one_bundle(**dss_event["match"])
     else:
         logger.error("Ignoring unknown event type %s", dss_event["event_type"])
+
+
+@contextmanager
+def db_writes():
```
Comment (kislyuk) [inline dcpquery/etl/__init__.py]: Is this necessary? What are the circumstances under which readonly_db starts out as True but you want to force it to False? Also, I'm not sure if it does what you expect. `readonly_db` is only in effect for remote db, and I think you may have to flush the database connection before and after changing it.

File: tests/unit/test_bundle_updates.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            f"SELECT COUNT(*) FROM bundle_file_links where bundle_fqid='{bundle_fqid}'").fetchall()[0][0] == 50
+        drop_one_bundle(bundle_uuid, bundle_version)
+        assert config.db_session.execute(
+            f"SELECT COUNT(*) FROM bundles WHERE fqid='{bundle_fqid}'").fetchall()[0][0] == 0
+
+        assert config.db_session.execute(
+            f"SELECT COUNT(*) FROM bundle_file_links where bundle_fqid='{bundle_fqid}'").fetchall()[0][0] == 0
+
+        post_deletion_file_count = config.db_session.execute("SELECT COUNT(*) from files;").fetchall()[0][0]
+        assert post_deletion_file_count < file_count
```
Comment (kislyuk) [inline tests/unit/test_bundle_updates.py]: Please rewrite the above asserts using unittest helper assertions, which provide more information when the assertion fails.

**Comment by kislyuk** [comment]
Also, we will need to integration test this, but that is blocked on rewiring the dev or integration deployment of query service to point to the corresponding DSS counterpart, not staging (so we can emulate a full bundle lifecycle by writing mock data in DSS). Can you file an issue to track that?

File: tests/unit/test_database.py [hunk 3]
```diff
@@ -369,7 +255,3 @@ def create_upsert_rules_in_db(cls):
     def create_recursive_functions_in_db(cls):
         config.db_session.execute(cls.get_all_children_function_sql + cls.get_all_parents_function_sql)
         config.db_session.commit()
-
-
-if __name__ == '__main__':
```
Comment (kislyuk) [inline tests/unit/test_database.py]: Why is this deleted? It is required to be able to run the file directly.

File: tests/unit/test_database_orm.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        self.assertEqual(result[0].uuid, vx_bundle.uuid)
+        self.assertEqual(result[0].version, vx_bundle.version)
+        expect_version = vx_bundle.version.strftime("%Y-%m-%dT%H%M%S.%fZ")
+        self.assertEqual(result[0].fqid, f"{vx_bundle.uuid}.{expect_version}")
+        self.assertDictEqual(result[0].aggregate_metadata, vx_bundle_aggregate_md)
+        self.assertEqual(len(result[0].files), len(vx_bundle.files))
+        self.assertSetEqual(set(f.fqid for f in vx_bundle.files), set(f.fqid for f in result[0].files))
+
+    def test_delete_bundles_deletes_list_of_bundles(self):
+        bundle_fqids = [bundle[0] for bundle in config.db_session.query("fqid FROM bundles LIMIT 3;").all()]
```
Comment (kislyuk) [inline tests/unit/test_database_orm.py]: Since this is the ORM test suite, it's probably best to replace all the `query()` queries with their ORM equivalents.

File: tests/unit/test_bundle_updates.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import unittest
+from unittest.mock import patch
+
+from dcpquery import config
+from dcpquery.etl import drop_one_bundle, process_bundle_event
+from tests import mock_bundle_deletion_event
+
+
+class BundleUpdateEvewnts(unittest.TestCase):
```
Comment (kislyuk) [inline tests/unit/test_bundle_updates.py]: s/BundleUpdateEvewnts/BundleUpdateEvents/

File: tests/unit/test_database_orm.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+from dcpquery.db import Bundle, BundleFileLink, Process, File
+from dcpquery.etl import load_links
+from tests import vx_bundle, vx_bf_links, vx_bundle_aggregate_md, mock_links
+
+from dcpquery import config
+
+
+class TestBundles(unittest.TestCase):
+    def test_insert_select_bundle(self):
+        # config.reset_db_session()
```
Comment (kislyuk) [inline tests/unit/test_database_orm.py]: Please uncomment or remove this line.

**Comment by kislyuk** [comment]
Tests are failing because the configuration changes in the new database tests now cause the async query tests to time out.

File: tests/unit/test_database.py (hunk 3 above)
Comment (MDunitz) [reply tests/unit/test_database.py]: 👍

Comment (MDunitz) [reply tests/unit/test_database.py]: I think it just ended up in a different file when I split the test files

File: tests/unit/test_database_orm.py (hunk 6 above)
Comment (MDunitz) [reply tests/unit/test_database_orm.py]: 👍

File: tests/unit/test_database_orm.py (hunk 4 above)
Comment (MDunitz) [reply tests/unit/test_database_orm.py]: I replaced the ones that have ORM equivalents but I only wrote functions for the most common, do you think its worth it to write them for all queries we are using?

File: tests/unit/test_bundle_updates.py (hunk 5 above)
Comment (MDunitz) [reply tests/unit/test_bundle_updates.py]: 👍

File: tests/unit/test_database_orm.py (hunk 4 above)
Comment (kislyuk) [reply tests/unit/test_database_orm.py]: I was referring to statements like `f"SELECT * FROM bundle_file_links where bundle_fqid='{bundle_fqid}'"`:
- This is a SQL injection vector. Even though the input is trusted here, as a matter of safety habit, never inline variables into SQL statements like this. Instead pass them to the executor as parameters.
- You can express this statement in ORM using something like `dcpquery.config.db_session.execute(select([BundleFileLink]).where(BundleFileLink.bundle_fqid == bundle_fqid)).fetchall()`

For any given test suite, I propose that we either use raw SQL (with parameters passed to psycopg2), or full-on ORM like the above. No need to rewrite existing statements now, but just the new ones you're adding.

Comment (kislyuk) [reply tests/unit/test_database_orm.py]: (You will need `from sqlalchemy.sql import select; from dcpquery.db import BundleFileLink` for the above to work)

**Comment by kislyuk** [comment]
The reason tests are failing on gitlab but not on your local workstation is that it takes longer for gitlab postgres to execute the following command:
```
  File "/HumanCellAtlas/query-service/tests/unit/test_database.py", line 44, in setUpClass
    config.db_session.execute("DROP RULE bundle_table_ignore_duplicate_inserts ON bundles;")
```
This previously reliably succeeded, I think because the database statement timeout that was being set in https://github.com/HumanCellAtlas/query-service/blob/master/tests/unit/test_create_async_query.py#L29 always ran first, initialized the db connection, and leaked state into the tests in `test_database`. We need to isolate the state by resetting the connection and set the timeout in `test_database` appropriately.

**Comment by kislyuk** [comment]
It also looks like one of the the new tests is not idempotent:
```
======================================================================
FAIL: test_drop_one_bundle_handles_deletion (tests.unit.test_bundle_updates.BundleUpdateEvents)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/akislyuk/projects/query-service/tests/unit/test_bundle_updates.py", line 19, in test_drop_one_bundle_handles_deletion
    f"SELECT COUNT(*) FROM bundles WHERE fqid='{bundle_fqid}'").fetchall()[0][0], 1)
AssertionError: 0 != 1

----------------------------------------------------------------------
```
I have to drop the database and reload the data manually before I can re-run the tests. Do you see this, and is it intentional?

**Comment by kislyuk** [comment]
Actually, it looks like my hypothesis about the timeout is wrong: I can reproduce the timeout locally by running `coverage run --source dcpquery -m unittest discover --start-directory tests/unit --top-level-directory . --verbose`, but it goes away when I run `python -m unittest discover --start-directory tests/unit --top-level-directory . --verbose` (no coverage tracking). It also goes away when running with coverage but disabling test_bundle_updates.py:BundleUpdateEvents.test_drop_one_bundle_handles_deletion. So that test is triggering an interaction between coverage tracking and something in the database driver.

**Comment by kislyuk** [comment]
OK, I worked around the coverage issue by switching it to a different mode of coverage tracing, and filed #214 to track down the root cause.

File: tests/unit/test_database_orm.py (hunk 4 above)
Comment (MDunitz) [reply tests/unit/test_database_orm.py]: 👍
